cmd/activator/main.go

```diff
@@ -236,7 +236,7 @@ func main() {
 
 	// NOTE: MetricHandler is being used as the outermost handler of the meaty bits. We're not interested in measuring
 	// the healthchecks or probes.
-	ah = activatorhandler.NewMetricHandler(env.PodName, ah)
+	ah = wrapActivatorHandlerWithFullDuplex(activatorhandler.NewMetricHandler(env.PodName, ah))
```

Comment:
Just for me to understand better, what is the reason that we need the handler twice?

Reply:
+1 

Is it not possible to do the annotation lookup and setting the duplex option when we proxy the request here ?

https://github.com/knative/serving/blob/75613869a913b7bc7e0b6615128eeaf7484877ff/pkg/activator/handler/handler.go#L126

Reply:
Secondly, I say we don't include this change on the metrics handler - I don't think users have reported an issue with that.

Reply:
Metrics handler is on the request path and activator logs the same error we are trying to address. If I don't wrap it I get that error,  not as frequently, but I do after a large number of requests.
Each handler interacts in isolation wrt other handlers in the chain when it comes to the full duplex thingy. Bottom line is that handlers that override write logic eg:
```
type Writer interface {
	Write(p []byte) (n int, err error)
}
```
create the issue seen. Probably we need an e2e test to copy the [test logic](https://github.com/skonto/test-reverse-proxy/tree/main#test-with-knative-serving) or you can [test this manually](https://github.com/skonto/test-reverse-proxy/tree/main#test-with-knative-serving).

Reply:
> Each handler interacts in isolation wrt other handlers in the chain, when it comes to the full duplex thingy. 

The underlying ResponseWriter is the one from net/http though - so EnableFullDuplex needs to only be called once.

https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/net/http/server.go;l=504-507

When we wrap this original writer in various handlers we should include an `Unwrap` method

https://cs.opensource.google/go/go/+/refs/tags/go1.21.4:src/net/http/responsecontroller.go;l=42-44

Reply:
It says that:
```
// The ResponseWriter should be the original value passed to the Handler.ServeHTTP method,
// or have an Unwrap method returning the original ResponseWriter.
```
The controller just needs the original writer so it can call the right methods including fullduplex.
Aren't we passing the original? I don't see creating another but maybe I miss something.

Reply:
I will check.

Reply:
So it seems you need to only set it at the outermost handler (first called). The unwrap thingy does not change the outcome but it is good practice so that the handlers can be used with a responsecontroller.